lib/keychain/keychain.rb

```diff
@@ -102,6 +136,37 @@ def generic_passwords
       Scope.new(Sec::Classes::GENERIC, self)
     end
 
+    # Imports item from string or file to this keychain
+    #
+    # @param [String, File] input String or File with raw data to import
+    # @param [Array <String>] trusted_apps List of applications which will be
+    # permitted to access imported items
+    def import(input, trusted_apps=[])
+      input = input.read if input.is_a? File
+
+      # Create <CFArray> with references to <SecTrustedApplication> objects
+      app_array = CF::Array.mutable.release_on_gc
+      trusted_apps.each do |app|
+        trusted_app = FFI::MemoryPointer.new(:pointer)
+        status = Sec.SecTrustedApplicationCreateFromPath(app.encode(Encoding::UTF_8), trusted_app)
```

Comment:
Yeah that stuff is a little tricky - 

trusted_app is just a pointer. You need to typecast to a CF::Base object to be able to release it, something like this:

```
trusted_app_buffer = FFI::MemoryPointer.new(:pointer)
status = Sec.SecTrustedApplicationCreateFromPath(app.encode(Encoding::UTF_8), trusted_app)

check_os_status(status)

trusted_app = CF::Base.typecast(trusted_app_buffer.read(:pointer)).release_on_gc

```
On 24 April 2015 at 17:18:02, Mikhail Zholobov (notifications@github.com) wrote:

In lib/keychain/keychain.rb:

> @@ -102,6 +136,37 @@ def generic_passwords
>        Scope.new(Sec::Classes::GENERIC, self)
>      end
>   
> +    # Imports item from string or file to this keychain
> +    #
> +    # @param [String, File] input String or File with raw data to import
> +    # @param [Array <String>] trusted_apps List of applications which will be
> +    # permitted to access imported items
> +    def import(input, trusted_apps=[])
> +      input = input.read if input.is_a? File
> +
> +      # Create <CFArray> with references to <SecTrustedApplication> objects
> +      app_array = CF::Array.mutable.release_on_gc
> +      trusted_apps.each do |app|
> +        trusted_app = FFI::MemoryPointer.new(:pointer)
> +        status = Sec.SecTrustedApplicationCreateFromPath(app.encode(Encoding::UTF_8), trusted_app)
Could you please give a hint - wow it can be fixed? Seems like release-methods are not supported for FFI::MemoryPointer instances

—
Reply to this email directly or view it on GitHub.